simulator/risc_v/riscv_decoder.h

```diff
@@ -58,29 +76,77 @@ struct RISCVInstrDecoder
             |  (J_imm20    << 20U);
     }
 
-    uint32 get_immediate_value( char subset) const noexcept
+    uint32 get_immediate_value( ImmediateType subset) const noexcept
     {
         switch (subset) {
-        case 'I': return I_imm;
-        case 'B': return get_B_immediate();
-        case 'S': return S_imm4_0 | (S_imm11_5 << 5);
-        case 'U': return U_imm;
-        case 'J': return get_J_immediate();
-        case 'C': return csr_imm;
-        case ' ': return 0;
-        default: assert(0); return 0;
+        case I: return I_imm;
+        case B: return get_B_immediate();
+        case S: return S_imm4_0 | (S_imm11_5 << 5U);
+        case U: return U_imm;
+        case J: return get_J_immediate();
+        case C: return csr_imm;
+        case NONE: return 0;
+        default: return get_compressed_immediate_value( subset);
+        }
+    }
+
+    uint16 get_compressed_immediate_value( ImmediateType type) const noexcept
```

Comment:
This function contains a lot of repetitive code. Please think about making it more readable.